ceno_emul/src/syscalls.rs

```diff
@@ -0,0 +1,79 @@
+use crate::{Change, EmuContext, VMState, WORD_SIZE, WordAddr, WriteOp};
+use anyhow::Result;
+use itertools::{Itertools, izip};
+use tiny_keccak::keccakf;
+
+/// A syscall event, available to the circuit witness generators.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct SyscallWitness {
+    pub mem_writes: Vec<WriteOp>,
+}
+
+/// The effects of a syscall to apply on the VM.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct SyscallEffects {
+    pub witness: SyscallWitness,
+    pub return_value: Option<u32>,
+    pub next_pc: Option<u32>,
+}
+
+pub const KECCAK_PERMUTE: u32 = 0x00_01_01_09;
+
+/// Trace the inputs and effects of a syscall.
+pub fn handle_syscall(vm: &VMState, function_code: u32, arg0: u32) -> Result<SyscallEffects> {
```

Comment:
Limiting ourselves to exactly one proper argument for each ecall seems like a weird limitation that we inherited from Risc0.

Reply:
Support for any number of registers was added in the mean time.